### PR TITLE
mysql: improve warnings when data directory is present.

### PR DESCRIPTION
**Pull request by peterb**
Users are already warned about config files, warn users about conflicting data directory too.

### CODE REVIEW

File: Formula/mysql@5.7.rb [hunk 1]
```diff
@@ -100,10 +100,17 @@ def caveats
...
+
+        [WARNING] /usr/local/var/mysql/ from another install may interfere with a Homebrew-built
+        server starting up correctly.
+      EOS
+    end
     if (my_cnf = ["/etc/my.cnf", "/etc/mysql/my.cnf"].find { |x| File.exist? x })
       s += <<~EOS
 
-        A "#{my_cnf}" from another install may interfere with a Homebrew-built
+        [WARNING] A "#{my_cnf}" from another install may interfere with a Homebrew-built
```
Comment (MikeMcQuaid) [inline Formula/mysql@5.7.rb:113]:
```suggestion
        A "#{my_cnf}" from another install may interfere with a Homebrew-built
```

File: Formula/mysql@5.7.rb [hunk 2]
```diff
@@ -100,10 +100,17 @@ def caveats
       To connect run:
           mysql -uroot
     EOS
+    if Dir.exist?("/usr/local/var/mysql/")
```
Comment (MikeMcQuaid) [inline Formula/mysql@5.7.rb:103]: Is it specifically `/usr/local/var/mysql/` or relative to the Homebrew prefix (i.e.): 

```suggestion
    if data = (var/"mysql") && data.exist?
```

File: Formula/mysql@5.7.rb [hunk 3]
```diff
@@ -100,10 +100,17 @@ def caveats
...
+      s += <<~EOS
+
+        [WARNING] /usr/local/var/mysql/ from another install may interfere with a Homebrew-built
```
Comment (MikeMcQuaid) [inline Formula/mysql@5.7.rb:106]:
```suggestion
        /usr/local/var/mysql/ from another install may interfere with a Homebrew-built
```

and perhaps

```suggestion
        #{var}/mysql from another install may interfere with a Homebrew-built
```

**Review by MikeMcQuaid — COMMENTED** [review]
Thanks for the PR! I think the `[WARNING]` is a bit out of place and a few questions/tweaks but otherwise: nice work!

**Comment by dtrodrigues** [comment]
Can you prefix the commit message with `mysql@5.7: ` please so it's clear that this is specific to mysql and not homebrew in general?

**Comment by peterb** [comment]
I've removed the warning, but this won't detect the presence of that data directory unless it can add something to the caveats from the `post_install` method, is this possible?

**Comment by MikeMcQuaid** [comment]
> I've removed the warning

Thanks!



> but this won't detect the presence of that data directory unless it can add something to the caveats from the `post_install` method, is this possible?

I'm not sure I understand this, can you explain why this would be useful?



> Update: I don't think that is possible, but at least users will get a note to remind them that this data directory could cause issues.

Yes, I was thinking this alone would be useful.

**Comment by peterb** [comment]
> I'm not sure I understand this, can you explain why this would be useful?

We could detect the presence of the data directory before the post_install runs and alter the caveats depending on the result, so the user could be reminded that they had already installed mysql there. I believe that when the user invokes the formula the install method runs, then the post_install method, and then the caveats method. When the program enters the post_install method, there isn't any way of modifying the caveats, but it's at this point when we could check to see if that data directory already exists and add a warning about it for the user to see. Once the program is in the caveats method, it's already too late to check if there was a pre-existing data directory as brew has already run the mysql initialize on it via the post_install method.

**Comment by MikeMcQuaid** [comment]
I see. In that case the change as-is seems rather pointless, unfortunately. Can you point to the specific issues that are being caused by this being already present and how failures occur?

**Comment by peterb** [comment]
Yes, I added some details to before I opened this issue on https://discourse.brew.sh/t/for-homebrew-mysql-installs-how-to-fix-mysql-sock-path/660/8

**Comment by MikeMcQuaid** [comment]
I see, thanks.

> Perhaps the install task should check if the data directory is present and abort if it is rather than silently failing?

I think in this case it needs to check specifically if the data directory exists _and_ is for a newer version of MySQL than the one you're installing. I'm not sure how to do this but it feels like the right direction (and would need to be made to `mysql.rb` and all `mysql@*.rb` formulae).

**Comment by gromgit** [comment]
@peterb You may want to look at how the `postgresql` formula deals with DB upgrades. In short, its caveat references an [external Homebrew command](https://docs.brew.sh/External-Commands) `postgresql-upgrade-database` that users can run to Do The Right Thing. The source is in `$(brew --repo homebrew/core)/cmd/postgresql-upgrade-database.rb`.

**Comment by MikeMcQuaid** [comment]
@gromgit This isn't necessary/comparable with MySQL because:
- MySQL will automatically upgrade databases without intervention whereas PostgreSQL requires a manual command to be run
- PostgreSQL uses separate data directories per-version and MySQL (how we have it setup, at least) does not. 

We could consider using separate data directories for the older versions of MySQL we have (but we'd need to have it only take effect if the datadir doesn't exist so we don't break existing installs).

**Comment by peterb** [comment]
@MikeMcQuaid I'm happy to look at the `mysql@*.rb` formulae so that they create and use separate data directories unless there's already one present for the specific version the user is installing. I'm assuming there is a way to detect which version of MySQL a data directory corresponds to, and that the `mysql` formula won't require a check as it'll auto-upgrade if there was data from an older version present?

**Comment by MikeMcQuaid** [comment]
> I'm assuming there is a way to detect which version of MySQL a data directory corresponds to

Yeh, there should be. I have no idea what it is, though 😭 



> that the `mysql` formula won't require a check as it'll auto-upgrade if there was data from an older version present?

Yes, that's a safe assumption 👍🏻

**Review by MikeMcQuaid — CHANGES_REQUESTED** [review]
Just so my review is here (as @SMillerDev requested this): as-is this unfortunately won't do anything but we could consider moving the `datadir` for new installs.

**Comment by peterb** [comment]
@MikeMcQuaid I'm looking into this already, have just pushed some of the changes.

**Comment by MikeMcQuaid** [comment]
Thanks @peterb, no rush!

File: Formula/mysql@5.6.rb [hunk 4]
```diff
@@ -22,7 +22,19 @@ class MysqlAT56 < Formula
...
+  end
+
+  # TODO: Could also try running file on files in var/*.frm if they exist
+  # e.g.
+  # peter@Peters-iMac homebrew-core % file /usr/local/var/mysql56/mysql/columns_priv.frm
+  # /usr/local/var/mysql56/mysql/columns_priv.frm: MySQL table definition file Version 9,
+  # type MYISAM, MySQL version 50647
+  #
+  def default_datadir_already_in_use_by_this_version?
+    (var/"mysql").exist? && var.glob("mysql/*.dblwr").empty? && (var/"mysql/test").exist?
```
Comment (MikeMcQuaid) [inline Formula/mysql@5.6.rb:37]: Think it's enough to just check for `(var/"mysql").exist?`?

File: Formula/mysql@5.7.rb [hunk 5]
```diff
@@ -149,7 +156,8 @@ def plist
 
     port = free_port
     pid = fork do
-      exec bin/"mysqld", "--bind-address=127.0.0.1", "--datadir=#{dir}", "--port=#{port}"
+      exec bin/"mysqld", "--bind-address=127.0.0.1", "--datadir=#{dir}", "--port=#{port}",
+      "--socket=#{dir}/mysql.sock}"
```
Comment (MikeMcQuaid) [inline Formula/mysql@5.7.rb:160]:
```suggestion
      "--socket=#{dir}/mysql.sock"
```

Any ideas why this is needed now but wasn't before?

File: Formula/mysql@5.6.rb (hunk 4 above)
Comment (MikeMcQuaid) [reply Formula/mysql@5.6.rb:37]: Could also output something in the `caveats` based on this?

File: Formula/mysql@5.7.rb (hunk 5 above)
Comment (peterb) [reply Formula/mysql@5.7.rb]: The CI build was failing on one of the target OSXs. I reproduced what I think is the same issue on my iMac by running `brew test mysql@5.7` without this line, and it fails and prints a message about the socket being taken. When I run it with the custom socket, it works.

File: Formula/mysql@5.6.rb (hunk 4 above)
Comment (peterb) [reply Formula/mysql@5.6.rb:37]: Sounds good,  I think that a simpler check and a caveat explaining the location of the datadir to the user would be preferable to trying to work it out automatically for them. I'll modify as you suggest.

Comment (peterb) [reply Formula/mysql@5.6.rb:37]: Update: It turns out this is harder to do than I expected, as putting a conditional in the `datadir` method means it gets evaluated differently during various stages in the formula.

File: Formula/mysql@5.7.rb (hunk 5 above)
Comment (peterb) [reply Formula/mysql@5.7.rb]: Update: Now that CI has run again, it seems that the socket file wasn't the issue. I haven't been able to reproduce this on my iMac, so will remove this custom socket.

**Comment by peterb** [comment]
@MikeMcQuaid I've updated this to something along the lines of what you were suggesting, I think that detecting if another version of MySQL is installed and using a different datadir is a nice idea in theory, but based on how the formula works it could make it more confusing for the user, and detecting the version of MySQL isn't going to be reliable because there are so many different heuristics but no definitive way of doing it.

With separate datadirs for each version (`mysql57` and `mysql56` are just suggestions), the possibilities of a conflict should be ameliorated quite a bit, as they can run `brew install mysql; brew install mysql@5.7` and use any of these versions by running `brew services stop mysql; brew services start msql@5.7` for people that have to deal with legacy databases, this should make using Homebrew to install them a lot friendlier.

_Summary:_ assuming a fresh OSX install, this would get rid of the gotcha that 5.7 will break and requires manually fixing the datadir if MySQL 8 is already installed. If they already have a legacy MySQL installed and decide to run reinstall, then the caveats will tell them that the datadir location has moved.

**Comment by peterb** [comment]
Not sure what the consensus is on this one, so closing it for the moment.

**Comment by MikeMcQuaid** [comment]
Reopening because it's on my list, sorry for the delay here.

**Comment by MikeMcQuaid** [comment]
@peterb https://github.com/Homebrew/homebrew-core/pull/65102 demonstrates the sort of approach I think would make sense to follow for MySQL. Let me know if you have any thoughts/questions but hopefully this will unblock you.

**Comment by peterb** [comment]
@MikeMcQuaid thanks! At a first glance #65102 looks like a good example, I'll take a more detailed look at it.

**Comment by peterb** [comment]
@MikeMcQuaid MySQL 5.6 has reached end-of-life, so it's probably only worth updating the formula for MySQL 5.7.

**Comment by peterb** [comment]
@MikeMcQuaid @SMillerDev @gromgit sorry for the delay in picking his up. I could look at MySQL 5.7? How valuable would be for Homebrew?

**Comment by SMillerDev** [comment]
Go for it, every improvement is valuable

**Pull request closed without merging**